File: applications/rtkadmmtotalvariation/rtkadmmtotalvariation.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkadmmtotalvariation_ggo.h"
#include "rtkGgoFunctions.h"
#include "rtkConfiguration.h"

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>

#include "rtkProjectionsReader.h"
#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkConstantImageSource.h"
#include "rtkADMMTotalVariationConeBeamReconstructionFilter.h"
#include "rtkPhaseGatingImageFilter.h"

int
main(int argc, char * argv[])
{
  GGO(rtkadmmtotalvariation, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
  using GradientOutputImageType = itk::CudaImage<itk::CovariantVector<OutputPixelType, Dimension>, Dimension>;
#else
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;
  using GradientOutputImageType = itk::Image<itk::CovariantVector<OutputPixelType, Dimension>, Dimension>;
#endif

  //////////////////////////////////////////////////////////////////////////////////////////
  // Read all the inputs
  //////////////////////////////////////////////////////////////////////////////////////////

  // Projections reader
  using projectionsReaderType = rtk::ProjectionsReader<OutputImageType>;
  auto projectionsReader = projectionsReaderType::New();
  rtk::SetProjectionsReaderFromGgo<projectionsReaderType, args_info_rtkadmmtotalvariation>(projectionsReader,
                                                                                           args_info);

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::endl;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));

  // Phase gating weights reader
  auto phaseGating = rtk::PhaseGatingImageFilter<OutputImageType>::New();
  if (args_info.phases_given)
  {
    phaseGating->SetPhasesFileName(args_info.phases_arg);
    phaseGating->SetGatingWindowWidth(args_info.windowwidth_arg);
    phaseGating->SetGatingWindowCenter(args_info.windowcenter_arg);
    phaseGating->SetGatingWindowShape(args_info.windowshape_arg);
    phaseGating->SetInputProjectionStack(projectionsReader->GetOutput());
    phaseGating->SetInputGeometry(geometry);
    TRY_AND_EXIT_ON_ITK_EXCEPTION(phaseGating->Update())
  }

  // Create input: either an existing volume read from a file or a blank image
  itk::ImageSource<OutputImageType>::Pointer inputFilter;
  if (args_info.input_given)
  {
    // Read an existing image to initialize the volume
    auto inputReader = itk::ImageFileReader<OutputImageType>::New();
    inputReader->SetFileName(args_info.input_arg);
    inputFilter = inputReader;
  }
  else
  {
    // Create new empty volume
    using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
    auto constantImageSource = ConstantImageSourceType::New();
    rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkadmmtotalvariation>(constantImageSource,
                                                                                                 args_info);
    inputFilter = constantImageSource;
  }

  //////////////////////////////////////////////////////////////////////////////////////////
  // Setup the ADMM filter and run it
  //////////////////////////////////////////////////////////////////////////////////////////

  // Set the reconstruction filter
  using ADMM_TV_FilterType =
    rtk::ADMMTotalVariationConeBeamReconstructionFilter<OutputImageType, GradientOutputImageType>;
  auto admmFilter = ADMM_TV_FilterType::New();

  // Set the forward and back projection filters to be used inside admmFilter
  SetForwardProjectionFromGgo(args_info, admmFilter.GetPointer());
  SetBackProjectionFromGgo(args_info, admmFilter.GetPointer());

  // Set all four numerical parameters
  admmFilter->SetCG_iterations(args_info.CGiter_arg);
  admmFilter->SetAL_iterations(args_info.niterations_arg);
  admmFilter->SetAlpha(args_info.alpha_arg);
  admmFilter->SetBeta(args_info.beta_arg);

  // Set the inputs of the ADMM filter
  admmFilter->SetInput(0, inputFilter->GetOutput());
  if (args_info.phases_given)
  {
    admmFilter->SetInput(1, phaseGating->GetOutput());
    admmFilter->SetGeometry(phaseGating->GetOutputGeometry());
    admmFilter->SetGatingWeights(phaseGating->GetGatingWeightsOnSelectedProjections());
  }
  else
  {
    admmFilter->SetInput(1, projectionsReader->GetOutput());
    admmFilter->SetGeometry(geometry);
  }
  admmFilter->SetDisableDisplacedDetectorFilter(args_info.nodisplaced_flag);

  REPORT_ITERATIONS(admmFilter, ADMM_TV_FilterType, OutputImageType)

  TRY_AND_EXIT_ON_ITK_EXCEPTION(admmFilter->Update())

  // Write the output
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(admmFilter->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkadmmtotalvariation/rtkadmmtotalvariation.py
#!/usr/bin/env python
import argparse
import sys
import itk
from itk import RTK as rtk


def build_parser():
    # Argument parsing
    parser = argparse.ArgumentParser(
        description="Perform ADMM total variation reconstruction on cone-beam projections.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )

    # General options
    parser.add_argument(
        "-v", "--verbose", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "-g", "--geometry", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "-o", "--output", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "-n", "--niterations", help="Number of iterations", type=int, default=1
    )
    parser.add_argument(
        "--alpha", help="Regularization parameter", type=float, default=0.1
    )
    parser.add_argument(
        "--beta",
        help="Augmented Lagrangian constraint multiplier",
        type=float,
        default=1.0,
    )
    parser.add_argument(
        "--CGiter",
        help="Number of nested iterations of conjugate gradient",
        type=int,
        default=5,
    )
    parser.add_argument("-i", "--input", help="Input volume", type=str)
    parser.add_argument(
        "--nodisplaced",
        help="Disable the displaced detector filter",
        action="store_true",
    )

    # Phase gating options
    parser.add_argument(
        "--phases", help="File containing the phase of each projection", type=str
    )
    parser.add_argument(
        "-c",
        "--windowcenter",
        help="Target reconstruction phase",
        type=float,
        default=0.0,
    )
    parser.add_argument(
        "-w",
        "--windowwidth",
        help="Tolerance around the target phase",
        type=float,
        default=1.0,
    )
    parser.add_argument(
        "-s",
        "--windowshape",
        help="Shape of the gating window",
        type=str,
        choices=["Rectangular", "Triangular"],
        default="Rectangular",
    )

    # RTK specific groups
    rtk.add_rtkprojectors_group(parser)
    rtk.add_rtkiterations_group(parser)
    rtk.add_rtkinputprojections_group(parser)
    rtk.add_rtk3Doutputimage_group(parser)

    # Parse the command line arguments
    return parser


def process(args_info: argparse.Namespace):

    # Define output pixel type and dimension
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Projections reader
    projectionsReader = rtk.ProjectionsReader[OutputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(projectionsReader, args_info)

    # Geometry
    if args_info.verbose:
        print(f"Reading geometry from {args_info.geometry}...")

    geometry = rtk.read_geometry(args_info.geometry)

    # Phase gating weights reader
    phaseGating = rtk.PhaseGatingImageFilter[OutputImageType].New()
    if args_info.phases:
        phaseGating.SetPhasesFileName(args_info.phases)
        phaseGating.SetGatingWindowWidth(args_info.windowwidth)
        phaseGating.SetGatingWindowCenter(args_info.windowcenter)
        # Rectangular=0, Triangular=1
        if args_info.windowshape == "Triangular":
            phaseGating.SetGatingWindowShape(1)
        else:
            phaseGating.SetGatingWindowShape(0)
        phaseGating.SetInputProjectionStack(projectionsReader.GetOutput())
        phaseGating.SetInputGeometry(geometry)

        phaseGating.Update()

    # Create input: either an existing volume read from a file or a blank image
    if args_info.input:
        # Read an existing image to initialize the volume
        inputFilter = itk.imread(args_info.input)
    else:
        # Create new empty volume
        constantImageSource = rtk.ConstantImageSource[OutputImageType].New()
        rtk.SetConstantImageSourceFromArgParse(constantImageSource, args_info)
        constantImageSource.Update()
        inputFilter = constantImageSource

    # Set the reconstruction filter
    if hasattr(itk, "CudaImage"):
        CudaOutputImageType = itk.CudaImage[OutputPixelType, Dimension]
        admmFilter = rtk.ADMMTotalVariationConeBeamReconstructionFilter[CudaOutputImageType, itk.CudaImage[itk.CovariantVector[OutputPixelType, Dimension], Dimension]].New()
        admmFilter.SetInput(0, itk.cuda_image_from_image(inputFilter.GetOutput()))
        if args_info.phases:
            admmFilter.SetInput(1, itk.cuda_image_from_image(phaseGating.GetOutput()))
        else:
            admmFilter.SetInput(
                1, itk.cuda_image_from_image(projectionsReader.GetOutput())
            )
    else:
        admmFilter = rtk.ADMMTotalVariationConeBeamReconstructionFilter[
            OutputImageType
        ].New()
        admmFilter.SetInput(0, inputFilter.GetOutput())
        if args_info.phases:
            admmFilter.SetInput(1, phaseGating.GetOutput())
        else:
            admmFilter.SetInput(1, projectionsReader.GetOutput())

    # Set the forward and back projection filters to be used inside admmFilter
    rtk.SetForwardProjectionFromArgParse(args_info, admmFilter)
    rtk.SetBackProjectionFromArgParse(args_info, admmFilter)

    # Set all four numerical parameters
    admmFilter.SetCG_iterations(args_info.CGiter)
    admmFilter.SetAL_iterations(args_info.niterations)
    admmFilter.SetAlpha(args_info.alpha)
    admmFilter.SetBeta(args_info.beta)

    admmFilter.SetDisableDisplacedDetectorFilter(args_info.nodisplaced)

    # Set the inputs of the ADMM filter
    if args_info.phases:
        admmFilter.SetGeometry(phaseGating.GetOutputGeometry())
        admmFilter.SetGatingWeights(phaseGating.GetGatingWeightsOnSelectedProjections())
    else:
        admmFilter.SetGeometry(geometry)

    rtk.SetIterationsReportFromArgParse(args_info, admmFilter)

    admmFilter.Update()

    itk.imwrite(admmFilter.GetOutput(), args_info.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
